src/backend/distributed/utils/maintenanced.c

```diff
@@ -257,6 +263,13 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		CHECK_FOR_INTERRUPTS();
 
 		/*
+		 * We clear the metadata cache before every iteration because otherwise
+		 * it might contain stale OIDs if deadlock detection runs during a DROP
+		 * EXTENSION followed by a CREATE EXTENSION.
+		 */
+		ClearMetadataCache();
```

Comment:
The problem is that `CitusHasBeenLoaded()` may get the old `pg_dist_partition` OID from `get_relname_relid` after it has already been invalidated. After that, no more invalidations for that OID arrive, meaning `InvalidateDistRelationCacheCallback` never clears the OID cache because it never matches the oid in `MetadataCache.distPartitionRelationId`.